### ISSUE TITLE
Documents removed after read time-out.

### ISSUE THREAD

**Issue by OkkeKlein** [Norconex/collector-http#98]
Is it possible to only remove documents with 404 status code?

(and also log the broken link)

**Comment by essiembre** [comment]
Maybe.. it depends whether you care about defining what are the valid/bad status code to delete or something else.  

**Deletion due to bad HTTP response status:**
If you only care to restrict what constitutes an invalid valid HTTP response code, you can specify this with this approach:
```xml
<documentFetcher>
    <validStatusCodes>200</validStatusCodes>
</documentFetcher>
```
The default is 200 (after redirect resolution). Anything but the list of HTTP status codes you provide (coma-separated) is considered "invalid" and will be deleted if it was previously added.

**Deletion due to other settings:**
In principle, it is not possible to prevent deletion of a document returning a bad processing status when it was added on a previous run.  For instance, this can happen when you define filter rules. 
Let say you have configured a filter to reject pages that have a specific strings in the content, or a metadata field that says: "expired" or something.  The first time you crawl, the page does not have that, so it is accepted and indexed.  The Nth time, the page status is "expired" and it is now rejected according to your filter.  So it needs to be "deleted" from your search engine (or else).  Otherwise, you'll end up with outdated/invalid content in your index. 

In either cases (HTTP status code or rules), if it was not added on a previous run, a rejected document will simply be ignored and will never reach your Committer (not added, not deleted).

Does that answer your question?

**Comment by OkkeKlein** [comment]
As mentioned in the title, I was more concerned with documents that got a read time-out when recrawling. As orphaned files are deleted this could lead to a lot of documents being removed from index. 

As I still want 404's to be logged and removed I can't use
```
<documentFetcher>
    <validStatusCodes>200</validStatusCodes>
</documentFetcher>
```

**Comment by essiembre** [comment]
I see your point. it will be picked up on next crawl, when the page starts responding properly again, thus avoiding unresponsive pages from showing up in your search results until they are OK again.

I understand there might be different cases where you would want to keep them in your index anyway.  So I am flagging this as a feature request to have more control about what triggers a deletion or not.

**Comment by essiembre** [comment]
It just occurred to me, about your request to log the broken links, you may be interested in this article: http://www.norconex.com/create-a-website-broken-links-checker/

**Comment by OkkeKlein** [comment]
Yeah I am using that one, but found out that if I only accept status code 200 in documentFetcher, I don't find the broken links anymore :)

**Comment by essiembre** [comment]
Please try the [latest snapshot](http://www.norconex.com/collectors/collector-http/download).  It has a new feature that may provide you what you need.  It now supports adding custom implementations of a new interface called ``ISpoiledReferenceStrategizer``.  A default implementation is provided called ``GenericSpoiledReferenceStrategizer``, where you can overwrite default handling of bad references (URLs) to get what you want.

See the documentation for it [here](http://www.norconex.com/collectors/collector-http/configuration) and [there](http://www.norconex.com/collectors/collector-core/latest/apidocs/com/norconex/collector/core/spoil/impl/GenericSpoiledReferenceStrategizer.html).

In addition, the [GenericDocumentFetcher](http://www.norconex.com/collectors/collector-http/latest/apidocs/com/norconex/collector/http/fetch/impl/GenericDocumentFetcher.html) class now allows you to define what constitutes a NOT_FOUND (if that ever matters to you).

Finally, when an HTTP document is fetched the crawler event listeners now receive as a "subject" a new class called HttpFetchResponse, which hold the exact HTTP status code and reason.

**Comment by OkkeKlein** [comment]
Great. Seems to be exactly what I'm looking for. WIll test it.

**Comment by OkkeKlein** [comment]
OK. As a test I dropped timeouts to 1 sec, so I would get "read timed out" and using the default settings  for fetcher (I didn't set anything in config) this resulted in deletion of all documents. (Solr couldn't actually handle the deletions. Will make another issue for this). As server didn't return a 404 I, did not expect this to happen. At the least I expected a GRACE_ONCE to be set.

Please advise.

**Comment by essiembre** [comment]
Works as expected when I test.  Did you set up ``spoiledStateStrategyResolver``? Can you past that config portion here?

If you just want the NOT_FOUND to be GRACE_ONCE and ensure the other ones are never deleted, you can do it like this:
```xml
<spoiledStateStrategyResolver 
    class="com.norconex.collector.core.state.impl.GenericSpoiledStateStrategyResolver"
    fallbackStrategy="INGORE">
  <mapping state="NOT_FOUND" strategy="GRACE_ONCE" />
  <mapping state="BAD_STATUS" strategy="IGNORE" />
  <mapping state="ERROR" strategy="IGNORE" />
</spoiledStateStrategyResolver>
```
That way timeouts should not generate deletions.

**Comment by OkkeKlein** [comment]
Dropping timeouts to 1 second to generate timeouts with
```<spoiledReferenceStrategizer 
      class="com.norconex.collector.core.spoil.impl.GenericSpoiledReferenceStrategizer"
      fallbackStrategy="IGNORE">
    <mapping state="NOT_FOUND" strategy="DELETE" />
	<mapping state="BAD_STATUS" strategy="IGNORE" />
	<mapping state="ERROR" strategy="IGNORE" />
    </spoiledReferenceStrategizer>
```

Results in 

website: 2015-07-15 16:22:17 ERROR - website: Could not process document: http://www.xxx (java.net.SocketTimeoutException: Read timed out)
com.norconex.collector.core.CollectorException: java.net.SocketTimeoutException: Read timed out

website: 2015-07-15 16:22:17 INFO - website: Deleting orphan references (if any)...

website: 2015-07-15 16:22:20 INFO -  DOCUMENT_COMMITTED_REMOV: ...........

So basically if read time outs occur the documents and subsequent orphaned references are removed. And not IGNORED

**Comment by essiembre** [comment]
I think I get what's happening now.  I see from your log that we are dealing with something different. To confirm whether a URL is deleted when it times out, you have to check if the URL that caused the timeout is the one being deleted.  I suspect that part works fine.

From your log snippet, it the deletion takes place when crawling is done and it deals with "orphans" (pages crawled in a previous run that could no longer be reached).  Have you set the orphan strategy to something other than IGNORE?  When set to be ignored, none of the orphans should be sent for deletion.

If you want to delete unreachable pages, a timeout or other temporary condition may prevent reaching a page and wrongfully mark it as unreachable.  This is a tricky situation.  My recommendation is to set the orphan strategy to PROCESS so unreachable URLs are crawled again.  If they are valid, they will remain in your index (maybe not a bad thing).  If they are not found, then your spoiled strategy should kick in.

If you really want to delete orphan URLs, no matter if they are still valid pages or not, then the timeout of parent pages is definitely an issue.  I am open to suggestions on how best to handle this corner case.

**Comment by OkkeKlein** [comment]
The document with the read timeout has crawlstate ERROR and isn't deleted as far as I can see. So then the references coming from this document are not orphans and should not be deleted.

**Comment by essiembre** [comment]
To resolve this specific issue I think we would need to keep breadcrumb trail of all pages crawled and their last status (all the parents that were crawled to get to a given page).  This is currently not possible and would require significant changes that would be too much for upcoming 2.2.0 release.  

In the meantime, to get around this issue, make sure you do not have DELETE an orphan strategy.  

Since this original question of this ticket is addressed (i.e. there IS a way to delete only 404), I will mark as fixed.  I created a new ticket to deal with deletion of "false-orphans" in a future release here: https://github.com/Norconex/collector-http/issues/126.

**Comment by essiembre** [comment]
Norconex HTTP Collector 2.2.0 official release is out. It includes this fix. You can download it [here](http://www.norconex.com/collectors/collector-http/download).
